### DISCUSSION TITLE
Chombo-style "incrementing" FluxRegisters?

### DISCUSSION THREAD

**Discussion by BenWibking** [AMReX-Codes/amrex#2140] [Q&A] [answered]
Hi,

I've written a single-level conservation law solver that allocates and stores fluxes grid-by-grid in `FArrayBox`es rather than in `MultiFab`s. From reading the example code in the Advection_AmrCore tutorial, it is not clear how to modify my code to do AMR with refluxing, since it the example appears to assume that I have a `MultiFab` for each flux component.

Is there a way to increment the flux registers grid-by-grid along the lines of what is possible in Chombo, using Chombo's `incrementCoarse` and `incrementFine` functions? Or do I need to re-write my code to use global MultiFabs for each flux component?

**Comment by WeiqunZhang** [comment, marked as answer]
You could use `YAFluxRegister` in `amrex/Src/Boundary`.  You can use `amrex/Tests/EB/CNS/` as an example.  The CNS code uses `EBFluxRegister`, which is derived from `YAFluxRegister`.  When you look at the CNS code, you only need to look at the block for `FabType::regular`.

**Reply by BenWibking** [reply]
Excellent! Thanks.

**Reply by BenWibking** [reply]
@WeiqunZhang Ok, so something is not quite working with my code when using YAFluxRegister, even without subcycling and only with a single level of refinement for a simple advection test. There are significant conservation errors at certain timesteps (changes of ~1e-5 in a single timestep). With max level set to 0, the conservation error is of order machine epsilon.

Since the API is different from EBFluxRegister, what exactly does YAFluxRegister expect for units/dimensions of the flux arrays? Do these need to be scaled in any way? My code uses a RK2-SSP integrator so it should be virtually identical to CNS::advance() except using YAFluxRegister instead of EBFluxRegister.

**Reply by WeiqunZhang** [reply]
EBFluxRegister adds some new functions, not really different API.  The EBCNS example code uses functions from both YAFluxRegister and EBFluxRegister.  Flux should not be scaled in any way by dt or area.

**Reply by WeiqunZhang** [reply]
For example, if we are talking about mass flux, then it should have the units of say `g/(s*cm^2)`

**Reply by WeiqunZhang** [reply]
I just converted the `Advection_AmrCore` tutorial to using `YAFluxRegister`, and put it here https://github.com/WeiqunZhang/amrex-devtests/tree/main/yafluxregister.  I have verified that the conservation error is at the machine precision level in both 2d and 3d, when `adv.do_reflux=1`.

**Reply by BenWibking** [reply]
I've created an example code that shows that amrex::FluxRegister and amrex::YAFluxRegister appear to give different answers. In this test problem, FluxRegister conserves mass at machine precision, but YAFluxRegister does not. So either I am using the flux register incorrectly, or there is a bug in the flux register.

I've made it so one can choose which FluxRegister to use at compile-time using ifdefs. I have copied your example code as closely as possible. The code is here: https://github.com/BenWibking/AMRAdvect. Would you be able to take a look at it? You can reproduce the conservation/lack of conservation by building and running `build/src/test_advection2d default.in`.

**Reply by WeiqunZhang** [reply]
Thanks for providing a reproducer.  The issue is we never considered that the blocking factor could be 1.  In your test, because the blocking factor is 1, there are a couple coarse cells with fine neighbors at both lower and upper faces of the same direction.  For example, at step 17, coarse cell (19,20)'s lo-y and hi-y neighbors are both fine cells.  We could fix this.  I will submit a PR soon.  However, for production runs, such a small blocking factor is **not** recommended.

**Reply by WeiqunZhang** [reply]
Please give #2155 a try.

**Reply by BenWibking** [reply]
It looks like that fixes it. However, this problem still occurs at least with a blocking factor of 2 (and perhaps even greater), since the connected components of the masked coarse grid (that is, when masked by the fine grid) do not have to obey the blocking factor. In the attached example, grids have a blocking factor of 2, but there are still coarse cells with fine neighbors at both lower and upper faces along the same direction due to the placement of the fine grids:
![AMR0001](https://user-images.githubusercontent.com/1781919/125150057-2b182c00-e180-11eb-9e7a-6c4a759af6ac.png)

**Reply by BenWibking** [reply]
Actually, no, there is still an issue. When I run the same problem but increase max_level to 2, I get a conservation error of ~1e-7 level from a single timestep. I've updated the repo to illustrate this issue. I've checked that this case also works when using FluxRegister.

**Reply by WeiqunZhang** [reply]
Yes, a blocking factor of 2 could also give rise to that kind of situation.  That is the gap between two fine grids are as small as 2 fine cells (1 coarse cell).  As for max_level=2, I will take a look at it this weekend.  But I wonder if the grids are properly nested.  Could you post a picture of the grids?  Did you run with a blocking factor of 1?  If so, could you also try with a blocking factor of 2?

**Reply by WeiqunZhang** [reply]
I tried your latest version with max_level=2.  It ran to step 73 and then hit a conservation error.  But I see a similar error with FluxRegister.  The issue is not FluxRegister or YAFluxRegister.  The issue is your state multifab has 4 ghost cells.  You have to use at least a blocking factor of 8, otherwise the ghost cells are not properly filled.  For testing, you can reduce the number of ghost cells to 1 and use a blocking factor of 2.

**Reply by BenWibking** [reply]
Ah, well, that would explain it. I can't use nghost < 3 due to the PPM reconstruction I am using.

Is this mentioned in the documentation anywhere? I did not see any note of this when reading the section on the gridding algorithm or the runtime parameters. This is a pretty significant caveat in choosing the blocking factor. Does the blocking factor have to be a multiple of nghost? Or just >= 2*nghost?

**Reply by BenWibking** [reply]
What causes this constraint? Does FillPatch only use the validbox on the coarse level?

**Reply by WeiqunZhang** [reply]
Blocking factor is usually better to be a power of 2.   For proper nesting, it needs to be greater than `(ceil(nghost,ref_ratio)+ninterp)*ref_ratio`, where `ninterp` (typically 1) depends on the scheme used for coarse/fine interpolation.  Yes this is because only coarse valid cells are used in interpolation.  Otherwise we would have to go down more levels, and that would not be very efficient.  However, amrex does support going down as many levels as needed in `Amr/AmrLevel`, although not in AmrCore.  There is also a check for proper nesting in `AmrLevel` (https://github.com/AMReX-Codes/amrex/blob/21.07/Src/Amr/AMReX_AmrLevel.cpp#L1030).  For `AmrCore` based code, you can check it yourself using https://github.com/AMReX-Codes/amrex/blob/21.07/Src/AmrCore/AMReX_FillPatchUtil_I.H#L9

Anyway, for performance, one should use a factor of at least 8.  For GPU, I would go with 16 or 32 at least

**Reply by BenWibking** [reply]
I've added a check for proper nesting. Thanks.

**Comment by WeiqunZhang** [comment]
> For proper nesting, it needs to be greater than `(ceil(nghost,ref_ratio)+ninterp)*ref_ratio`, where `ninterp` (typically 1) depends on the scheme used for coarse/fine interpolation.

It should be greater than or equal to `(ceil(nghost/ref_ratio)+ninterp)*ref_ratio`.
